### ISSUE TITLE
Optimize: Matmul+Transpose Fusion

### ISSUE THREAD

**Issue by hikettei** [hikettei/Caten#109]
- [ ] !reshape always use (!contiguous x)
- [ ] In a certain condition, !contiguous wont create an additional tensor.

**Comment by hikettei** [comment]
as well as transposed gemm

**Comment by hikettei** [comment]
- Pattern Matcher based solution is enough good, implement:
  - [ ] matmul.transpose_x/y
  - [ ] TransposeNode (as a module)

**Comment by hikettei** [comment]
todo: writing a detail

**Comment by hikettei** [comment]
Currently, `A @ B.T` produces two kernels.

```lisp
CATEN/TEST-SUITE>
(ctx:with-contextvar (:jit 1 :jit_debug 1 :packed 0)
  (caten (!matmul (make-tensor `(3 5)) (!t (make-tensor `(3 5))))))
Compiled[e6]:

/*
Arrays:
  - val_3[float32]: (5 3) // OUTPUT, TMP
  - val_0[float32]: (5 3) // INPUT, TMP
*/
void main4507277_e6_k0(float* val_3, float* val_0);
void main4507277_e6_k0(float* val_3, float* val_0) {
  for(int _gid0=0;(_gid0<=4);_gid0+=1) {
    for(int _gid1=0;(_gid1<=2);_gid1+=1) {
      val_3[((3*_gid0)+_gid1)] = val_0[(_gid0+(5*_gid1))];
    }
  }
}

/*
Arrays:
  - val_16[float32]: (3 3 1) // OUTPUT, TMP
  - val_3[float32]: (1 3 5) // INPUT, TMP
  - val_9[float32]: (3 1 5) // INPUT, TMP
*/
void main4507277_e6_k1(float* val_16, float* val_3, float* val_9);
void main4507277_e6_k1(float* val_16, float* val_3, float* val_9) {
  for(int _gid0=0;(_gid0<=2);_gid0+=1) {
    for(int _gid1=0;(_gid1<=2);_gid1+=1) {
      val_16[((3*_gid0)+_gid1)] = 0.0;
      for(int _gid2=0;(_gid2<=4);_gid2+=1) {
        val_16[((3*_gid0)+_gid1)] += val_9[((5*_gid0)+_gid2)] * val_3[(_gid1+(3*_gid2))];
      }
    }
  }
}
#S(AVM
   :GRAPH 
Graph[seen=NIL, outputs=NIL] {
    <ALLOCATE : val_9 <- (shape=(3, 5), stride=(5, 1)) where :nrank=2 :dtype=FLOAT32 :_type_relay=<OK>>
    <ALLOCATE : val_0 <- (shape=(3, 5), stride=(5, 1)) where :nrank=2 :dtype=FLOAT32 :_type_relay=<OK>>
    <ALLOCATE : val_3 <- (shape=(1, 3, 5), stride=(15, 1, 3)) where :nrank=3 :dtype=FLOAT32 :_type_relay=<OK>>
    <Node[JIT] JIT_KERNEL(NID4558098) : val_3, val_0 <- (val_3, val_0) where :fname=MAIN4507277_e6_k0 :jit-info=<CLANG[NIL] Code [2 kernels]>>
    <ALLOCATE : val_16 <- (shape=(3, 3, 1), stride=(3, 1, 1)) where :nrank=3 :dtype=FLOAT32 :_type_relay=<OK>>
    <Node[JIT] JIT_KERNEL(NID4558099) : val_16, val_3, val_9 <- (val_16, val_3, val_9) where :fname=MAIN4507277_e6_k1 :jit-info=<CLANG[NIL] Code [2 kernels]>>
    <Node[SPECIAL/VM] PAUSE/BACKWARD(NID4558017) : val_21 <- (val_16) where :_type_relay=<OK> :_no_group_realize_on_vm=T>
}

   :NAME :MAIN4507277
   :FW-OUTPUTS (|val_21|)
   :BW-OUTPUTS NIL
   :ID2TENSOR #<HASH-TABLE :TEST EQL :COUNT 1 {703DD254C3}>
   :TAPE-LENGTH 7
   :PC 0
   :VARIABLES #<HASH-TABLE :TEST EQL :COUNT 2 {703DF143E3}>
   :DUMPED NIL)
CATEN/TEST-SUITE> 
```

**Comment by hikettei** [comment]
Solved at #138
